dm/master/scheduler/scheduler.go

```diff
@@ -0,0 +1,1139 @@
+	s.wg.Add(2)
+	go func() {
+		defer func() {
+			s.wg.Done()
+			close(workerEvCh)
+			close(workerErrCh)
+		}()
+		ha.WatchWorkerEvent(ctx, etcdCli, rev+1, workerEvCh, workerErrCh)
+	}()
+	go func() {
+		defer s.wg.Done()
+		s.handleWorkerEv(ctx, workerEvCh, workerErrCh)
+	}()
+
+	s.started = true // started now
+	s.cancel = cancel
+	s.etcdCli = etcdCli
+	s.logger.Info("the scheduler has started")
+	return nil
+}
+
+// Close closes the scheduler.
+func (s *Scheduler) Close() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.started {
+		return
+	}
+
+	s.logger.Info("the scheduler is closing")
+	if s.cancel != nil {
+		s.cancel()
+		s.cancel = nil
+	}
+
+	s.wg.Wait()
+	s.started = false // closed now.
+	s.logger.Info("the scheduler has closed")
+}
+
+// AddSourceCfg adds the upstream source config to the cluster.
+// NOTE: please verify the config before call this.
+func (s *Scheduler) AddSourceCfg(cfg config.MysqlConfig) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.started {
+		return terror.ErrSchedulerNotStarted.Generate()
+	}
+
+	// 1. check whether exists.
+	if _, ok := s.sourceCfgs[cfg.SourceID]; ok {
+		return terror.ErrSchedulerSourceCfgExist.Generate(cfg.SourceID)
+	}
+
+	// 2. put the config into etcd.
+	_, err := ha.PutSourceCfg(s.etcdCli, cfg)
+	if err != nil {
+		return err
+	}
+
+	// 3. record the config in the scheduler.
+	s.sourceCfgs[cfg.SourceID] = cfg
+
+	// 4. try to bound it to a Free worker.
+	bounded, err := s.tryBoundForSource(cfg.SourceID)
+	if err != nil {
+		return err
+	} else if !bounded {
+		// 5. record the source as unbounded.
+		s.unbounds[cfg.SourceID] = struct{}{}
+	}
+	return nil
+}
+
+// RemoveSourceCfg removes the upstream source config in the cluster.
+// when removing the upstream source config, it should also remove:
+// - any existing relay stage.
+// - any source-worker bound relationship.
+func (s *Scheduler) RemoveSourceCfg(source string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.started {
+		return terror.ErrSchedulerNotStarted.Generate()
+	}
+
+	// 1. check whether the config exists.
+	_, ok := s.sourceCfgs[source]
+	if !ok {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(source)
+	}
+
+	// 2. check whether any subtask exists for the source.
+	existingSubtasksM := make(map[string]struct{})
+	for task, cfg := range s.subTaskCfgs {
+		for source2 := range cfg {
+			if source2 == source {
+				existingSubtasksM[task] = struct{}{}
+			}
+		}
+	}
+	existingSubtasks := strMapToSlice(existingSubtasksM)
+	if len(existingSubtasks) > 0 {
+		return terror.ErrSchedulerSourceOpTaskExist.Generate(source, existingSubtasks)
+	}
+
+	// 3. find worker name by source ID.
+	var (
+		workerName string // empty should be fine below.
+		worker     *Worker
+	)
+	if w, ok := s.bounds[source]; ok {
+		worker = w
+		workerName = w.BaseInfo().Name
+	}
+
+	// 4. delete the info in etcd.
+	_, err := ha.DeleteSourceCfgRelayStageSourceBound(s.etcdCli, source, workerName)
```

Comment:
what if source is unbounded? which means workerName=""

Reply:
`clientv3.OpDelete(common.UpstreamBoundWorkerKeyAdapter.Encode(worker))` works fine for `workerName=""` because no `clientv3.WithPrefix()` used.

And, I have already added a comment above
```go
workerName string // empty should be fine below.
```

Reply:
sorry, I haven't noticed that before